app/test/unit/git/for-each-ref-test.ts

```diff
@@ -58,5 +59,13 @@ describe('git/for-each-ref', () => {
       const branches = await getBranches(repo)
       expect(branches.length).to.equal(0)
     })
+
+    it('should return empty list for directory without a .git directory', async () => {
+      const emptyDir = mkdirSync('no-repo-here')
```

Comment:
I would prefer this setup be done in a `before` block instead of in the `it` test case

Reply:
This is likely a side effect of how our tests have built up over time, but I've favoured `before(Each)` and `after(Each)` setup and teardown when it's applied to a group of tests to eliminate the repetition. 

For a single test like this, it feels like a lot of ceremony - particularly with how mocha structures things (and TS deals with things defined outside scope). I think I'd have to have something like this:

```ts
describe('directory without a .git directory', function() {
  // because I need to set and read this from two different functions
  // this signature is required and makes me feel bad
  let repository: Repository | null = null

  before(function() {
    const emptyDir = mkDirSync('no-repo-here')
    repository = new Repository(emptyDir, -1, null, false)
  });

  it('should return empty list', async () => {
      // gotta ! the value here to tell tsc that it's fiiiiine
      const status = await getRemotes(repository!)
      expect(status).eql([])
  })
});
```

For a single test like this, I might split it up into [AAA](http://wiki.c2.com/?ArrangeActAssert) groupings if there's a lot of setup required, but even that is a flag for a potential code smell.

Reply:
that's fair! (i tend to prefer the `before` even for single test cases because it more clearly delineates setup from test but no big deal in this case)

Reply:
Yeah, if you feel like grabbing some existing tests and giving them some love we can keep this discussion moving.